File: packages/java/src/main/java/com/mattstrick/arrayfirst/ArrayFirst.java
package com.mattstrick.arrayfirst;

import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * Get the first element or first n elements of a list. Java port of
 * https://github.com/mattstrick/array-first-javascript
 */
public final class ArrayFirst {

    private ArrayFirst() {}

    /** Returns the first element, or empty Optional if the list is null or empty. */
    public static <T> Optional<T> first(List<T> list) {
        if (list == null || list.isEmpty()) return Optional.empty();
        return Optional.of(list.get(0));
    }

    /**
     * Returns the first n elements. If n is 1, returns a list of one element.
     * If n < 1, treated as 1.
     */
    public static <T> List<T> first(List<T> list, int n) {
        if (list == null || list.isEmpty()) return Collections.emptyList();
        int take = n < 1 ? 1 : Math.min(n, list.size());
        return list.stream().limit(take).collect(Collectors.toList());
    }
}

File: packages/kotlin/src/main/kotlin/com/mattstrick/arrayfirst/ArrayFirst.kt
package com.mattstrick.arrayfirst

/**
 * array-first (Kotlin)
 *
 * Get the first element or first n elements of a list.
 * Kotlin port of https://github.com/mattstrick/array-first-javascript
 *
 * License: MIT
 */

/**
 * Returns the first element of the list, or `null` if the list is empty.
 */
fun <T> first(list: List<T>): T? = list.firstOrNull()

/**
 * Returns the first [n] elements of the list.
 * When n is 1, returns a single-element list (use [first(List)] for a single element as [T]?).
 *
 * @param list the source list
 * @param n number of elements (must be >= 1)
 * @return list of the first n elements (may have fewer if list size < n)
 * @throws IllegalArgumentException if n < 1
 */
fun <T> first(list: List<T>, n: Int): List<T> {
    require(n >= 1) { "array-first expects n >= 1" }
    return list.take(n)
}
